### PR TITLE
Add logResources option

### LINKED ISSUE #142: Create a log/audit of all resources discovered, and whether they had tags etc
Would be more complete than the activity-audit log.

### PR DESCRIPTION
**Pull request by alutman-innablr**
https://github.com/Innablr/revolver/issues/142

* Changed the schema for `saveResources`. There's now a different field to control how resources are outputted at the start of execution.
* Added `console` and `csv` options to `resourceLog`
* Updated `audit.console` to use the logger rather than console.log
* TODO later, add `tags()` method to `toolingInterface` so the console and CSV can print tags.

Example
```
defaults:
  settings:
      resourceLog:
        json:
          file: resources.json
        console:
          reportTags: ["Name"]
        csv:
          file: resources.csv
          reportTags: ["Name"]
```

### CODE REVIEW

**Comment by simon-anz** [comment]
The names "audit" and "resourceLog" feel inconsistent.  Should the former be "auditLog"?

**Comment by simon-anz** [comment]
Please include the account-name (from config), value of the availabilityTag (in that account, if overridden), and Name Tag.

File: lib/resourceLog.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  abstract process(): void;
+}
+
+export class ResourceLogConsole extends ResourceLog {
+  constructor(entries: ToolingInterface[]) {
+    super(entries);
+  }
+  process(): void {
+    const header = `${'ACCOUNT_ID'.padEnd(16)}${'REGION'.padEnd(20)}${'TYPE'.padEnd(20)} ${'ID'.padEnd(40)} ${'STATE'}`;
+    const lines = this.entries.map((r) => `${(r.accountId || '').padEnd(16)}${(r.region || '').padEnd(20)}${(r.resourceType || '').padEnd(20)} ${r.resourceId.padEnd(40)} ${r.resourceState}`)
```
Comment (simon-anz) [inline lib/resourceLog.ts]: Can we do this (and the other similar lines) in a slightly more structured way, that's easier to view and edit.  Maybe a list of string+pad, then a loop/function that pads the values and applies the separators.

File: lib/resourceLog.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+export class ResourceLogCsv extends ResourceLog {
+  private readonly outputFile;
+  constructor(entries: ToolingInterface[], outputFile: string) {
+    super(entries);
+    this.outputFile = outputFile;
+  }
+  process(): void {
+    const header = `${'ACCOUNT_ID'.padEnd(16)}${'REGION'.padEnd(20)}${'TYPE'.padEnd(20)} ${'ID'.padEnd(40)} ${'STATE'}`;
+    const lines = this.entries.map((r) => `${(r.accountId || '').padEnd(16)}${(r.region || '').padEnd(20)}${(r.resourceType || '').padEnd(20)} ${r.resourceId.padEnd(40)} ${r.resourceState}`)
```
Comment (simon-anz) [inline lib/resourceLog.ts]: "CSV" is not C

File: lib/resourceLog.ts (hunk 1 above)
Comment (simon-anz) [reply lib/resourceLog.ts]: Maybe

```typescript
    const lines = this.entries.map((r) => [
      (r.accountId || '').padEnd(16),
      (r.region || '').padEnd(20),
      (r.resourceType || '').padEnd(20),
      r.resourceId.padEnd(40),
      r.resourceState
    ].join(''));
```

Maybe it's not worth the bother??

**Comment by alutman-innablr** [comment]
> The names "audit" and "resourceLog" feel inconsistent. Should the former be "auditLog"?

Changed.

> Please include the account-name (from config), value of the availabilityTag (in that account, if overridden), and Name Tag.

Added the account name but the availability tag is a plugin setting. We'd either need a pretty hacky condition to look at all the plugins to see if there's one with an availabilityTag field or restructure how the resource log executes.
If it's important, I can add the hacky lookup.

**Comment by simon-anz** [comment]
I think it's worthwhile adding this, as it is most-common use-case.  
You could push the responsibility for selecting "extra tags" up to the caller (AccountRevolver), and pass an `IncludeTags` parameter. The caller can look in powercycle+powercycleCentral (if enabled), and select their `.availabilityTag`

**Comment by alutman-innablr** [comment]
Added option to CSV and console to specify tags to report. Linking this value to the plugin config is a lot more pain (plugins can have several configs with different tags)

**Review by simon-anz — APPROVED** [review]
LGTM. Sample output:

image

**Pull request merged**